### ISSUE TITLE
Multi-file checking with no file passed in at cmd

### ISSUE THREAD

**Issue by DeeeeLAN** [SublimeLinter/SublimeLinter#1604]
Due to the nature of my project, I am unable to compile single files. Referencing #1528 and #1540, I updated my linter regex to use `?P<filename>` instead of `?P<file>`, and that sort of worked, but there were still issues because the file was still being passed in directly. 

I found #1520 and swapped the ${file} in my cmd for `${fake_file}`, which corrected the command that gets run to what I am expecting. However, now my current file does not receive any warnings or errors. The linter console states 'No lint results.' The debug output does show the warnings and errors resulting from the command. 

Is something else required to make this work?

**Comment by braver** [comment]
SublimeLinter is file oriented because it primarily wants to help you while you’re editing. Some linters like tslint for TypeScript will lint the file in the context of a project. But even then, linting is initiated from a file and the linter program looks for the parent project, and results are reported back for that file (not for the project as a while). There may be ways to hook it up, but it’s kinda outside SublimeLinter’s comfort zone, so we should first establish that’s what you need.

It sounds like what you’re doing is not “linting” so much as running a compiler and wanting to get in-editor feedback about errors. Have you tried using Sublime’s  built in “build system”? https://www.sublimetext.com/docs/3/build_systems.html

**Comment by kaste** [comment]
Well, this should be supported as of now. It probably comes down to what you capture as filename. Note that we throw away all errors coming from files currently not open. 

Since this is a new feature there can be bugs.

**Comment by DeeeeLAN** [comment]
The current output has the full system path, not just the base name of the file. Would this cause issues?

**Comment by kaste** [comment]
Absolute paths should be 👌.

Show your linter and the debug output?

**Comment by kaste** [comment]
Also maybe just try #1591 which has the better filename handling already.

**Comment by DeeeeLAN** [comment]
@braver, I am indeed looking to run the compiler and get in-editor feedback about warnings and errors. Unfortunately, this particular language (SystemVerilog) does not work well when running against individual files, because any dependencies within a file will cause a cascade of errors if the dependencies are not compiled simultaneously. However, the current flow works because I do only care about errors pertaining to the current file - saving the file, causing a recompile, then pulling only the relevant errors works great. When I have gotten it to work, it works great. Something is preventing it from working currently, however. I have used build system, but it outputs results in the console, instead of inline. Is there a plugin or a method to configure build system for inline reporting of warnings and errors? 

@kaste, the linter is [SublimeLinter-contrib-modelsim](https://github.com/jevogel/SublimeLinter-contrib-modelsim/blob/master/linter.py) (the vlog one), however I have modified it a bit, as shown here:

**Comment by kaste** [comment]
Well, sure the `/./` could be funky. Now, you got the cmd right and it hopefully prints the expected output, but even in debug mode we don't log how we parsed that output telling you that the regex is right. 

You could look at it by implementing `split_match`:

```
def split_match(self, match):
	error = super().split_match(match)
	print(error)
	return error
```

You could also just git clone the repo into your packages folder and instrument the `group_by_filename_and_update` function https://github.com/SublimeLinter/SublimeLinter/blob/master/sublime_linter.py#L331
. E.g. print the arguments it receives. 

Ensure, that you don't fall off this `if` statement https://github.com/SublimeLinter/SublimeLinter/blob/master/sublime_linter.py#L364-L365
bc then the captured filename does not match an open file.

**Comment by DeeeeLAN** [comment]
Thanks, that helped significantly. I was able to root-cause it - it is indeed failing because `window.find_open_file(filename)` fails. It is happening because I have a symlink in my path. Sublime is using the actual file paths for each open file, but my build results are using the symlinked path. 

If I change line 364 to `view = window.find_open_file(os.path.realpath(filename))`, the error shows up.

Since sublime uses the realpath for each tab anyway, I bet you could implement this change without breaking any setup. Of course, that can't be guaranteed, so maybe it would take a preference toggle: `resolve_symlinks = true`, or something.

**Comment by kaste** [comment]
Using `realpath` in an UI is problematic. If the user edits `/foo/bar.js` which in "reality" is at `/some/where/deep/foo/bar.js` we expect the linter to report errors on the short, user path. "Resolving" the path would actually break here for the majority of users. 

Does your linter produce the same output if used on the command line? (Honestly, the linter output is not really human readable, or at least a good UI, so they could improve this a lot.) 

Is the `/./` actually redundant here so that a `normpath` would suffice?

**Comment by DeeeeLAN** [comment]
`/./` was ignored, that had no effect on the situation. Yes, the command produces the same output at the command line: both use the symbolic root path, while sublime uses the actual root path. I do see how that could cause issues though, since some build environments may rely on such behavior. I now think a preferences toggle would be an ideal solution.

**Comment by kaste** [comment]
So I would recommend you implement `split_match` in your plugin. E.g.

```
def split_match(self, match):
	error = super().split_match(match)
	error['filename'] = os.path.normpath(error['filename'])
	return error
```

I would also open an issue against the linter because what they report as filenames isn't really user friendly. It looks like they just do a `join(working_dir, relative_filepath)` which is a bit sloppy.

**Comment by DeeeeLAN** [comment]
Thanks, that worked great. I still had to use `os.path.realpath` in this case due to the symbolic links, instead of `normpath`, but other than that, it is running exactly as expected. I will definitely open an issue with the linter owners, thanks for the suggestion.
